File: src/main/kotlin/com/huawei/cangjie/descriptors/impl/AbstractClassDescriptor.java
package com.huawei.cangjie.descriptors.impl;

import com.huawei.cangjie.descriptors.*;
import com.huawei.cangjie.incremental.components.NoLookupLocation;
import com.huawei.cangjie.name.Name;
import com.huawei.cangjie.psi.CjFile;
import com.huawei.cangjie.resolve.DescriptorUtils;
import com.huawei.cangjie.resolve.descriptorUtil.DescriptorUtilsKt;
import com.huawei.cangjie.resolve.lazy.descriptors.LazyExtendClassDescriptor;
import com.huawei.cangjie.resolve.scopes.*;
import com.huawei.cangjie.storage.NotNullLazyValue;
import com.huawei.cangjie.storage.StorageManager;
import com.huawei.cangjie.types.*;
import com.huawei.cangjie.types.checker.CangJieTypeRefiner;
import com.huawei.cangjie.types.util.TypeUtils;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiFile;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class AbstractClassDescriptor extends ModuleAwareClassDescriptor {
    protected final NotNullLazyValue<SimpleType> defaultType;
    private final Name name;
    private final NotNullLazyValue<MemberScope> unsubstitutedInnerClassesScope;
    private final NotNullLazyValue<ReceiverParameterDescriptor> thisAsReceiverParameter;

    private final StorageManager storageManager;
    private Set<LazyExtendClassDescriptor> extendClassDescriptor = null;


    public AbstractClassDescriptor(@NotNull StorageManager storageManager, @NotNull Name name) {
        this.storageManager = storageManager;
        this.name = name;
        this.defaultType = storageManager.createLazyValue(new Function0<>() {
            @Override
            public SimpleType invoke() {

                return TypeUtils.makeUnsubstitutedType(
                        AbstractClassDescriptor.this, getUnsubstitutedMemberScope(),
                        new Function1<>() {
                            @Override
                            public SimpleType invoke(CangJieTypeRefiner cangjieTypeRefiner) {
                                ClassifierDescriptor descriptor = cangjieTypeRefiner.refineDescriptor(AbstractClassDescriptor.this);
                                // If we've refined descriptor
                                if (descriptor == null) return defaultType.invoke();

                                if (descriptor instanceof TypeAliasDescriptor) {
                                    return CangJieTypeFactory.computeExpandedType(
                                            (TypeAliasDescriptor) descriptor,
                                            TypeUtils.getDefaultTypeProjections(descriptor.getTypeConstructor().getParameters())
                                    );
                                }

                                if (descriptor instanceof ModuleAwareClassDescriptor) {
                                    TypeConstructor refinedConstructor = descriptor.getTypeConstructor().refine(cangjieTypeRefiner);
                                    return TypeUtils.makeUnsubstitutedType(
                                            refinedConstructor,
                                            ((ModuleAwareClassDescriptor) descriptor).getUnsubstitutedMemberScope(cangjieTypeRefiner),
                                            this
                                    );
                                }

                                return descriptor.getDefaultType();
                            }
                        }
                );
            }
        });
        this.unsubstitutedInnerClassesScope = storageManager.createLazyValue(new Function0<MemberScope>() {
            @Override
            public MemberScope invoke() {
                return new InnerClassesScopeWrapper(getUnsubstitutedMemberScope());
            }
        });
        this.thisAsReceiverParameter = storageManager.createLazyValue(() -> new LazyClassReceiverParameterDescriptor(AbstractClassDescriptor.this));


    }

    @NotNull
    @Override
    public ReceiverParameterDescriptor getThisAsReceiverParameter() {
        return thisAsReceiverParameter.invoke();
    }

    @NotNull
    @Override
    public Name getName() {
        return name;
    }

    @NotNull
    @Override
    public List<ReceiverParameterDescriptor> getContextReceivers() {
        return Collections.emptyList();
    }

    @Override
    public @NotNull DescriptorVisibility getVisibility() {
        return DescriptorVisibilities.PUBLIC;

    }

    @NotNull
    @Override
    public MemberScope getUnsubstitutedMemberScope() {
        return getUnsubstitutedMemberScope(DescriptorUtilsKt.getCangJieTypeRefiner(DescriptorUtils.getContainingModule(this)));
    }

    public Set<LazyExtendClassDescriptor> getExtendClassDescriptors() {
        if (extendClassDescriptor == null) {
            getExtendClass();
        }
        return extendClassDescriptor;
    }

    /**
     * 获取扩展类型  （权宜之计）
     *
     * @return
     */
    public Set<LazyExtendClassDescriptor> getExtendClass() {
        extendClassDescriptor = new HashSet<>(ScopeUtilsKt.getExtendClasss(getCurrentEditorScope(), name, NoLookupLocation.FROM_PACKAGE));

        return extendClassDescriptor;
    }

    /**
     * 获取相对于活动编辑的 LexicalScope   （权宜之计）
     */
    public LexicalScope getCurrentEditorScope() {
        FileEditorManager fileEditorManager = FileEditorManager.getInstance(storageManager.getProject());
        Editor editor = fileEditorManager.getSelectedTextEditor();
        PsiFile file = editor == null ? null : PsiDocumentManager.getInstance(storageManager.getProject()).getPsiFile(editor.getDocument());
        if (file instanceof CjFile) {
            return ScopeUtilsKt.getScope((CjFile) file);
        }
        return null;
    }


    @NotNull
    @Override
    public MemberScope getMemberScope(@NotNull List<? extends TypeProjection> typeArguments, @NotNull CangJieTypeRefiner cangjieTypeRefiner) {
        assert typeArguments.size() == getTypeConstructor().getParameters().size() : "Illegal number of type arguments: expected "
                + getTypeConstructor().getParameters().size() + " but was " + typeArguments.size()
                + " for " + getTypeConstructor() + " " + getTypeConstructor().getParameters();
        if (typeArguments.isEmpty()) return getUnsubstitutedMemberScope(cangjieTypeRefiner);

        TypeSubstitutor substitutor = TypeConstructorSubstitution.create(getTypeConstructor(), typeArguments).buildSubstitutor();
        return new SubstitutingScope(getUnsubstitutedMemberScope(cangjieTypeRefiner), substitutor);
    }

    @NotNull
    @Override
    public MemberScope getMemberScope(@NotNull TypeSubstitution typeSubstitution, @NotNull CangJieTypeRefiner cangjieTypeRefiner) {
        if (typeSubstitution.isEmpty()) return getUnsubstitutedMemberScope(cangjieTypeRefiner);

        TypeSubstitutor substitutor = TypeSubstitutor.create(typeSubstitution);
        return new SubstitutingScope(getUnsubstitutedMemberScope(cangjieTypeRefiner), substitutor);
    }

    @NotNull
    @Override
    public MemberScope getMemberScope(@NotNull List<? extends TypeProjection> typeArguments) {
        return getMemberScope(typeArguments, DescriptorUtilsKt.getCangJieTypeRefiner(DescriptorUtils.getContainingModule(this)));
    }

    @NotNull
    @Override
    public MemberScope getMemberScope(@NotNull TypeSubstitution typeSubstitution) {
        return getMemberScope(typeSubstitution, DescriptorUtilsKt.getCangJieTypeRefiner(DescriptorUtils.getContainingModule(this)));
    }


    @NotNull
    @Override
    public ClassDescriptor getOriginal() {
        return this;
    }

    @NotNull
    @Override
    public MemberScope getUnsubstitutedInnerClassesScope() {
//        throw new UnsupportedOperationException("Should not be called on " + getClass());
        return unsubstitutedInnerClassesScope.invoke();
    }


    //
    @NotNull
    @Override
    public ClassDescriptor substitute(@NotNull TypeSubstitutor substitutor) {
        if (substitutor.isEmpty()) {
            return this;
        }
        return new LazySubstitutingClassDescriptor(this, substitutor);
    }

    @NotNull
    @Override
    public SimpleType getDefaultType() {
//        throw new UnsupportedOperationException("Should not be called on " + getClass());
        return defaultType.invoke();

    }

    @Override
    public void acceptVoid(DeclarationDescriptorVisitor<Void, Void> visitor) {
        visitor.visitClassDescriptor(this, null);
    }

    @Override
    public <R, D> R accept(DeclarationDescriptorVisitor<R, D> visitor, D data) {
        return visitor.visitClassDescriptor(this, data);
    }

    @Nullable
    @Override
    public SimpleType getDefaultFunctionTypeForSamInterface() {
        return null;
    }

    @Override
    public boolean isDefinitelyNotSamInterface() {
        return false;
    }

}

File: src/main/kotlin/com/huawei/cangjie/descriptors/impl/AbstractClassDescriptor.kt
//package com.huawei.cangjie.descriptors.impl
//
//import com.huawei.cangjie.descriptors.*
//import com.huawei.cangjie.name.Name
//import com.huawei.cangjie.resolve.DescriptorUtils.getContainingModule
//import com.huawei.cangjie.resolve.descriptorUtil.getCangJieTypeRefiner
//import com.huawei.cangjie.resolve.scopes.InnerClassesScopeWrapper
//import com.huawei.cangjie.resolve.scopes.MemberScope
//import com.huawei.cangjie.resolve.scopes.SubstitutingScope
//import com.huawei.cangjie.storage.NotNullLazyValue
//import com.huawei.cangjie.storage.StorageManager
//import com.huawei.cangjie.types.*
//import com.huawei.cangjie.types.CangJieTypeFactory.computeExpandedType
//import com.huawei.cangjie.types.TypeConstructorSubstitution.Companion.create
//import com.huawei.cangjie.types.TypeSubstitutor.Companion.create
//import com.huawei.cangjie.types.checker.CangJieTypeRefiner
//import com.huawei.cangjie.types.util.TypeUtils
//import com.huawei.cangjie.types.util.TypeUtils.getDefaultTypeProjections
//import com.huawei.cangjie.types.util.TypeUtils.makeUnsubstitutedType
//
//@OptIn(TypeRefinement::class)
//abstract class AbstractClassDescriptor(storageManager: StorageManager, override val name: Name) :
//    ModuleAwareClassDescriptor() {
//    protected val defaultType: NotNullLazyValue<SimpleType>
//    private val unsubstitutedInnerClassesScope: NotNullLazyValue<MemberScope>
//    private val thisAsReceiverParameter: NotNullLazyValue<ReceiverParameterDescriptor>
//
//    init {
//        this.defaultType = storageManager.createLazyValue {
//            makeUnsubstitutedType(
//                this@AbstractClassDescriptor, unsubstitutedMemberScope
//            ) { cangjieTypeRefiner: CangJieTypeRefiner? ->
//                val descriptor =
//                    cangjieTypeRefiner!!.refineDescriptor(this@AbstractClassDescriptor)
//                        ?: return@makeUnsubstitutedType defaultType.invoke()
//                // If we've refined descriptor
//
//                if (descriptor is TypeAliasDescriptor) {
//                    return@makeUnsubstitutedType descriptor.computeExpandedType(
//                        getDefaultTypeProjections(
//                            descriptor.getTypeConstructor().parameters
//                        )
//                    )
//                }
//
//                if (descriptor is ModuleAwareClassDescriptor) {
//                    val refinedConstructor =
//                        descriptor.getTypeConstructor().refine(cangjieTypeRefiner)
//                    return@makeUnsubstitutedType    makeUnsubstitutedType(
//                        refinedConstructor,
//                        descriptor.getUnsubstitutedMemberScope(
//                            cangjieTypeRefiner
//                        ),
//                        this
//                    )
//                }
//                descriptor.defaultType
//            }
//        }
//        this.unsubstitutedInnerClassesScope = storageManager.createLazyValue<MemberScope> {
//            InnerClassesScopeWrapper(
//                unsubstitutedMemberScope
//            )
//        }
//        this.thisAsReceiverParameter = storageManager.createLazyValue<ReceiverParameterDescriptor> {
//            LazyClassReceiverParameterDescriptor(this@AbstractClassDescriptor)
//        }
//    }
//
//    override fun getThisAsReceiverParameter(): ReceiverParameterDescriptor {
//        return thisAsReceiverParameter.invoke()
//    }
//
//    override fun getContextReceivers(): List<ReceiverParameterDescriptor> {
//        return emptyList()
//    }
//
//    override val visibility: DescriptorVisibility
//        get() = DescriptorVisibilities.PUBLIC
//
//    override fun getUnsubstitutedMemberScope(): MemberScope {
//        return getUnsubstitutedMemberScope(getContainingModule(this).getCangJieTypeRefiner())
//    }
//
//    override fun getMemberScope(
//        typeArguments: List<TypeProjection>,
//        cangjieTypeRefiner: CangJieTypeRefiner
//    ): MemberScope {
//        assert(typeArguments.size == typeConstructor.parameters.size) {
//            ("Illegal number of type arguments: expected "
//                    + typeConstructor.parameters.size + " but was " + typeArguments.size
//                    + " for " + typeConstructor + " " + typeConstructor.parameters)
//        }
//        if (typeArguments.isEmpty()) return getUnsubstitutedMemberScope(cangjieTypeRefiner)
//
//        val substitutor = create(
//            typeConstructor, typeArguments
//        ).buildSubstitutor()
//        return SubstitutingScope(getUnsubstitutedMemberScope(cangjieTypeRefiner), substitutor)
//    }
//
//
//
//    override fun getMemberScope(
//        typeSubstitution: TypeSubstitution,
//        cangjieTypeRefiner: CangJieTypeRefiner
//    ): MemberScope {
//        if (typeSubstitution.isEmpty()) return getUnsubstitutedMemberScope(cangjieTypeRefiner)
//
//        val substitutor = create(typeSubstitution)
//        return SubstitutingScope(getUnsubstitutedMemberScope(cangjieTypeRefiner), substitutor)
//    }
//
//    override fun getMemberScope(typeArguments: List<TypeProjection>): MemberScope {
//        return getMemberScope(typeArguments, getContainingModule(this).getCangJieTypeRefiner())
//    }
//
//    override fun getMemberScope(typeSubstitution: TypeSubstitution): MemberScope {
//        return getMemberScope(typeSubstitution, getContainingModule(this).getCangJieTypeRefiner())
//    }
//
//
//    override val original: ClassDescriptor
//        get() = this
//
//    override fun getUnsubstitutedInnerClassesScope(): MemberScope {
////        throw new UnsupportedOperationException("Should not be called on " + getClass());
//        return unsubstitutedInnerClassesScope.invoke()
//    }
//
//    //    @NotNull
//    //    @Override
//    //    public ReceiverParameterDescriptor getThisAsReceiverParameter() {
//    //        return thisAsReceiverParameter.invoke();
//    //    }
//    //
//    //    @NotNull
//    //    @Override
//    //    public List<ReceiverParameterDescriptor> getContextReceivers() {
//    //        return Collections.emptyList();
//    //    }
//    //
//    //    @NotNull
//    //    @Override
//    //    public MemberScope getMemberScope(@NotNull List<? extends TypeProjection> typeArguments, @NotNull CangJieTypeRefiner kotlinTypeRefiner) {
//    //        assert typeArguments.size() == getTypeConstructor().getParameters().size() : "Illegal number of type arguments: expected "
//    //                + getTypeConstructor().getParameters().size() + " but was " + typeArguments.size()
//    //                + " for " + getTypeConstructor() + " " + getTypeConstructor().getParameters();
//    //        if (typeArguments.isEmpty()) return getUnsubstitutedMemberScope(kotlinTypeRefiner);
//    //
//    //        TypeSubstitutor substitutor = TypeConstructorSubstitution.create(getTypeConstructor(), typeArguments).buildSubstitutor();
//    //        return new SubstitutingScope(getUnsubstitutedMemberScope(kotlinTypeRefiner), substitutor);
//    //    }
//    //
//    //    @NotNull
//    //    @Override
//    //    public MemberScope getMemberScope(@NotNull TypeSubstitution typeSubstitution, @NotNull CangJieTypeRefiner kotlinTypeRefiner) {
//    //        if (typeSubstitution.isEmpty()) return getUnsubstitutedMemberScope(kotlinTypeRefiner);
//    //
//    //        TypeSubstitutor substitutor = TypeSubstitutor.create(typeSubstitution);
//    //
//    //        return new SubstitutingScope(getUnsubstitutedMemberScope(kotlinTypeRefiner), substitutor);
//    //    }
//    //    @NotNull
//    //    @Override
//    //    public MemberScope getMemberScope(@NotNull List<? extends TypeProjection> typeArguments) {
//    //        return getMemberScope(typeArguments, DescriptorUtilsKt.getCangJieTypeRefiner(DescriptorUtils.getContainingModule(this)));
//    //    }
//    //    @NotNull
//    //    @Override
//    //    public MemberScope getMemberScope(@NotNull TypeSubstitution typeSubstitution) {
//    //        return getMemberScope(typeSubstitution, DescriptorUtilsKt.getCangJieTypeRefiner(DescriptorUtils.getContainingModule(this)));
//    //    }
//    //    @NotNull
//    //    @Override
//    //    public MemberScope getUnsubstitutedMemberScope() {
//    //        return getUnsubstitutedMemberScope(DescriptorUtilsKt.getCangJieTypeRefiner(DescriptorUtils.getContainingModule(this)));
//    //    }
//    //
//    override fun substitute(substitutor: TypeSubstitutor): ClassDescriptor {
//        if (substitutor.isEmpty) {
//            return this
//        }
//        return LazySubstitutingClassDescriptor(this, substitutor)
//    }
//
//    override fun getDefaultType(): SimpleType {
////        throw new UnsupportedOperationException("Should not be called on " + getClass());
//        return defaultType.invoke()
//    }
//
//    override fun acceptVoid(visitor: DeclarationDescriptorVisitor<Void, Void>) {
//        visitor.visitClassDescriptor(this, null)
//    }
//
//    override fun <R, D> accept(visitor: DeclarationDescriptorVisitor<R, D>, data: D?): R? {
//        return visitor.visitClassDescriptor(this, data)
//    }
//
//    override fun getDefaultFunctionTypeForSamInterface(): SimpleType? {
//        return null
//    }
//
//    override fun isDefinitelyNotSamInterface(): Boolean {
//        return false
//    }
//}
